apps/emqx_durable_storage/src/emqx_ds_storage_bitfield_lts.erl

```diff
@@ -0,0 +1,423 @@
+    lists:foreach(
+        fun(Msg) ->
+            {Key, _} = make_key(S, Msg),
+            Val = serialize(Msg),
+            rocksdb:put(DB, Data, Key, Val, [])
+        end,
+        Messages
+    ).
+
+get_streams(_Shard, #s{trie = Trie}, TopicFilter, _StartTime) ->
+    Indexes = emqx_ds_lts:match_topics(Trie, TopicFilter),
+    [#stream{storage_key = I} || I <- Indexes].
+
+make_iterator(_Shard, _Data, #stream{storage_key = StorageKey}, TopicFilter, StartTime) ->
+    %% Note: it's a good idea to keep the iterator structure lean,
+    %% since it can be stored on a remote node that could update its
+    %% code independently from us.
+    {ok, #it{
+        topic_filter = TopicFilter,
+        start_time = StartTime,
+        storage_key = StorageKey
+    }}.
+
+next(_Shard, Schema = #s{ts_offset = TSOffset}, It, BatchSize) ->
+    %% Compute safe cutoff time.
+    %% It's the point in time where the last complete epoch ends, so we need to know
+    %% the current time to compute it.
+    Now = emqx_message:timestamp_now(),
+    SafeCutoffTime = (Now bsr TSOffset) bsl TSOffset,
+    next_until(Schema, It, SafeCutoffTime, BatchSize).
+
+next_until(_Schema, It, SafeCutoffTime, _BatchSize) when It#it.start_time >= SafeCutoffTime ->
+    %% We're in the middle of the current epoch, so we can't yet iterate over it.
+    %% It would be unsafe otherwise: messages can be stored in the current epoch
+    %% concurrently with iterating over it. They can end up earlier (in the iteration
+    %% order) due to the nature of keymapping, potentially causing us to miss them.
+    {ok, It, []};
+next_until(#s{db = DB, data = CF, keymappers = Keymappers}, It, SafeCutoffTime, BatchSize) ->
+    #it{
+        start_time = StartTime,
+        storage_key = {TopicIndex, Varying}
+    } = It,
+    %% Make filter:
+    Inequations = [
+        {'=', TopicIndex},
+        {StartTime, '..', SafeCutoffTime - 1},
+        %% Unique integer:
+        any
+        %% Varying topic levels:
+        | lists:map(
+            fun
+                ('+') ->
+                    any;
+                (TopicLevel) when is_binary(TopicLevel) ->
+                    {'=', hash_topic_level(TopicLevel)}
+            end,
+            Varying
+        )
+    ],
+    %% Obtain a keymapper for the current number of varying levels.
+    NVarying = length(Varying),
+    %% Assert:
+    NVarying =< ?WILDCARD_LIMIT orelse
+        error({too_many_varying_topic_levels, NVarying}),
+    Keymapper = array:get(NVarying, Keymappers),
+    Filter =
+        #filter{range_min = LowerBound, range_max = UpperBound} = emqx_ds_bitmask_keymapper:make_filter(
+            Keymapper, Inequations
+        ),
+    {ok, ITHandle} = rocksdb:iterator(DB, CF, [
+        {iterate_lower_bound, emqx_ds_bitmask_keymapper:key_to_bitstring(Keymapper, LowerBound)},
+        {iterate_upper_bound, emqx_ds_bitmask_keymapper:key_to_bitstring(Keymapper, UpperBound + 1)}
+    ]),
+    try
+        put(?COUNTER, 0),
```

Comment:
This counter is apparently unused?

Reply:
True. But I think it can be used later for performance monitoring (to find out how well the indexing work, and how often it misses the bitmask).